### PR TITLE
Cancel streamed downloads & fix progress for downloads

### PR DESCRIPTION
**Pull request by kuhnroyal**
### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### CODE REVIEW

File: dio/test/cancel_token_test.dart
Comment (kuhnroyal) [inline dio/test/cancel_token_test.dart:44]: This can probably be reduced to 500 or even less.

**Comment by AlexV525** [comment]
Before review: Could we abstract the implementation of this cancellation above the specific adapter? If we're doing this in adapters, then every adapter needs to consider the same functionality.

**Comment by kuhnroyal** [comment]
I will see what I can do, but I already know that this is basically not supported on web.

**Comment by kuhnroyal** [comment]
> Before review: Could we abstract the implementation of this cancellation above the specific adapter? If we're doing this in adapters, then every adapter needs to consider the same functionality.

This is hard without breaking. If we want to do this in the `DioMixin` we have no access to the underlying http implementation and can not close the stream/connection. We would have to pass some new argument to the adapters or extend `ResponseBody` with a new `close` function. But this seems like an issue for dio 6.

**Comment by kuhnroyal** [comment]
This is a lot more complex. We have completely different behavior for different adapters when doing streamed responses/downloads and most of it can not be abstracted away.

For `IOHttpClientAdapter` we have a `ResponseBody` with a stream so the `await dio.get` succeeds after the connection is established with a `Response` and the cancellation exception is thrown on the stream.

For `BrowserHttpClientAdapter` we have a `ResponseBody` with bytes and the adapter has to wait until everything is received, before completing with a response. So the cancellation exception is thrown on the `await dio.get`.
And there is no way to access other information from the XHR when aborting.

**Comment by kuhnroyal** [comment]
> What about the HTTP/2 adapter?

Fixed it there as well. I think we should merge this and brain storm if we can abstract this away separately.

**Review by AlexV525 — APPROVED** [review]
LGTM

File: plugins/http2_adapter/lib/src/http2_adapter.dart
Comment (AlexV525) [inline plugins/http2_adapter/lib/src/http2_adapter.dart:158]: We have asynchronous gaps between this call and other awaits. Can we do this earlier or separated?

**Comment by AlexV525** [comment]
Is this still on your track? The last time I review this it only remains a few nits.

**Comment by AlexV525** [comment]
@kuhnroyal Is this still on your track? The last time I review this it only remains a few nits.

**Comment by kuhnroyal** [comment]
> @kuhnroyal Is this still on your track? The last time I review this it only remains a few nits.

Yea

**Comment by kuhnroyal** [comment]
Superseeded by https://github.com/cfug/dio/pull/2068 due to lots of updates on main

**Pull request closed without merging**
